### ISSUE TITLE
overtraining at the end of epoch when using large queue sizes

### ISSUE THREAD

**Issue by emilbols** [DL4Jets/DeepJetCore#30]
This appears to be unrelated to issue #28 so im opening a new issue.

When batch_size*max_queue_size is of the order of the dataset (~1/2 of the dataset), you get some kind of overtraining at the end of the epochs.

![maxqsize_2_0_5_lossperbatch](https://user-images.githubusercontent.com/33452471/64340394-565eec80-cfe6-11e9-935c-f6bca3cfc595.png)

I couldnt find what causes this, but just be aware when training on smaller datasets to not use large queue sizes.

**Comment by jkiesele** [comment]
Thank you for reporting this!
Does this still persist when you switch of the threaded reading of the data?

**Comment by emilbols** [comment]
yes, it is there in both cases

**Comment by emilbols** [comment]
So i tried to start a new training of DF to check if i got the same performance, and it still didn't look quite right.

It seems that this problems actually occurs even when the dataset is huge, its just harder to spot. Here is the batch loss zooming in on the end of an epoch.

![FullKeras205LargeQsize_lossperbatch](https://user-images.githubusercontent.com/33452471/64515369-c45c2a00-d2ec-11e9-9e80-fd6f610ac9a1.png)

The epoch ends a bit after batch 128250. It looks like a small effect, but it seems to hurt the performance quite a bit. Maybe because it is at the end of the epoch, right before models are saved or evaluated for the val_loss. I tried comparing q_size =  50 to q_size = 5. The small q-size is still training, but it seems to perform quite a bit better. This difference in val_loss is significant btw.

![loss](https://user-images.githubusercontent.com/33452471/64515488-f66d8c00-d2ec-11e9-9555-7216273a2942.png)

It might be it is also affecting the max_q_size = 5 to a lesser degree. This is the batch loss for that training. Epoch ends 59200.

![FullKeras200SmallQsize_lossperbatch](https://user-images.githubusercontent.com/33452471/64516060-0f2a7180-d2ee-11e9-9585-244498dd5513.png)

**Comment by jkiesele** [comment]
Thanks again for the investigations.
I am afraid we are at a point where the issue needs to be somehow brought to keras.
In that context, probably a simple one-file check is needed, without DJC dependencies.
At least this check might help to pinpoint if it is indeed keras only or some interplay problem.

**Comment by emilbols** [comment]
I would assume this is a DJC problem, whereas the exploding batches are connected with Keras. When i trained, deepflavour in the past i was using a low max_q_size, so it could be that this has always been here. Anyways i will check this issue more carefully later this week.

**Comment by jkiesele** [comment]
In the meantime, could you make a PR adding a (default False) option to training_base to read the full dataset in memory? (like --in_memory) or something similar?
(I can also do it, but since you brought it up...)

**Comment by emilbols** [comment]
so printing the batches from within the loss and comparing with the batches yielded from the generator, reveals that things seems to work fine, until the generator reaches the final batch of the epoch, and the queue appears to be dropped and that final batch gets shown repeatedly until the epoch ends. i might check with a simple generator to see if this always happens, but for now i think i will just max_q_size = 1.

and yes i can make the pr

**Comment by jkiesele** [comment]
Thanks!
Is the final batch repeatedly being provided by the generator in DeepJetCore or is this coming from keras? In the latter case, we can just adjust the DJC queueing.

**Comment by emilbols** [comment]
it gets shown once by the generator, which then starts working on the queue for the next epoch. However keras drops everything the generator prepared in the queue, and just starts showing that last batch

So its coming from keras

**Comment by jkiesele** [comment]
Ok, once your PR (full dataset in memory) is through, I will add a fix for the generator queue such that the DJC buffers are used, but not the keras ones.
Thanks again for checking all this.

**Comment by mverzett** [comment]
@emilbols, I would suggest escalating this to keras as well, it might be almost invisible in the DL community due to their very small batches, but indeed it is there.

**Comment by jkiesele** [comment]
For now, I have added some fixes here and there:
https://github.com/DL4Jets/DeepJetCore/commit/ef981fe5b2a6bca12a037021dc78b0a80614992d
and parent.
I will NOT update the package list here, because if usually comes with headaches, but the public installation on cern afs is downgraded to keras 2.0.5 accordingly.
This means, for now we lose multi-gpu support from keras. If the issue takes longer in keras, we might implement it back by hand.

**Comment by jkiesele** [comment]
I'll leave this one open as a reminder that the current fixes are just supposed to be temporary.
(he said... then, one year later ;) )

**Comment by jkiesele** [comment]
Actually, this has been fully resolved with the new underlying data structure
